**Цель работы:**

Осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных.

# Загрузка набора данных

### Описание предметной области

Вариант №15

Набор данных: clients.csv

Атрибуты:
ID-Уникальный идентификатор клиента

Year_Birth- Год рождения клиента

Education-Уровень образования клиента

Marital_Status-Семейное положение клиента

Income -Годовой доход семьи

Kidhome- количество детей

Dt_Customer- дата регистрации клиента в компании

NumDealsPurchases- количество покупок

### 1.Чтение файла (набора данных)

Google Colab работает в облаке на серверах Google. Когда нужно запустить блокнот, он изначально не имеет доступа к файлам на нашем жестком диске. Этот код создает кнопку для выбора и загрузки файлов.

In [2]:
# импорт библиотек, чтение файла с помощью pandas
import pandas as pd
df=pd.read_csv('clients.csv',sep=';')

Импорт библиотеки и чтение таблицы

### 2. Обзор данных

2.1 Вывод первых 20 строк с помощью метода head.

In [3]:
# применить метод head
df.head(20)

ID  Year_Birth   Education Marital_Status   Income  Kidhome Dt_Customer  \
0   5524        1957  Graduation         Single  58138.0      0.0  04.09.2012   
1   2174        1954  Graduation         Single  46344.0      1.0  08.03.2014   
2   4141        1965  Graduation       Together  71613.0      0.0  21.08.2013   
3   6182        1984  Graduation       Together  26646.0      1.0  10.02.2014   
4   5324        1981         PhD        Married  58293.0      1.0  19.01.2014   
5   7446        1967      Master       Together  62513.0      0.0  09.09.2013   
6    965        1971  Graduation       Divorced  55635.0      0.0  13.11.2012   
7   6177        1985         PhD        Married  33454.0      1.0  08.05.2013   
8   4855        1974         PhD       Together  30351.0      1.0  06.06.2013   
9   5899        1950         PhD       Together   5648.0      1.0  13.03.2014   
10  1994        1983  Graduation        MARRIED      NaN      NaN         NaN   
11   387        1976       Basic        Married   7500.0      0.0  13.11.2012   
12  2125        1959  Graduation       Divorced  63033.0      0.0  15.11.2013   
13  8180        1952      Master       Divorced  59354.0      1.0  15.11.2013   
14  2569        1987  Graduation        Married  17323.0      0.0  10.10.2012   
15  2114        1946         PhD          SINGL  82800.0      0.0  24.11.2012   
16  9736        1980  Graduation        Married  41850.0      1.0  24.12.2012   
17  4939        1946  Graduation       Together  37760.0      0.0  31.08.2012   
18  6565        1949      Master        Married  76995.0      0.0  28.03.2013   
19  9360        1982  Graduation        Married  37040.0      0.0  08.08.2012   

    NumDealsPurchases  
0                 3.0  
1                 2.0  
2                 1.0  
3                 2.0  
4                 5.0  
5                 2.0  
6                 4.0  
7                 2.0  
8                 1.0  
9                 1.0  
10                NaN  
11                1.0  
12                1.0  
13                3.0  
14                1.0  
15                1.0  
16                3.0  
17                2.0  
18                2.0  
19                1.0

Возвращает первые 20 строк таблицы

2.2 Оценка данных с помощью метода info.

In [4]:
# выполнит метод info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 796 entries, 0 to 795
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   ID                 796 non-null    int64  
 1   Year_Birth         796 non-null    int64  
 2   Education          796 non-null    object 
 3   Marital_Status     796 non-null    object 
 4   Income             784 non-null    float64
 5   Kidhome            795 non-null    float64
 6   Dt_Customer        795 non-null    object 
 7   NumDealsPurchases  795 non-null    float64
dtypes: float64(3), int64(2), object(3)
memory usage: 49.9+ KB


Используется метод info, который предоставляет общую информацию о таблице, включая названия столбцов, типы данных и количество ненулевых значений в каждом столбце

2.3 Оценка данных с помощью метода describe.

In [5]:
# оцените числовые столбцы с помощью describe
df.describe()

ID   Year_Birth        Income     Kidhome  NumDealsPurchases
count    796.000000   796.000000     784.00000  795.000000         795.000000
mean    5630.133166  1968.356784   53130.07398    0.438994           2.314465
std     3273.039715    12.022132   21818.56876    0.547252           1.941650
min        0.000000  1899.000000    2447.00000    0.000000           0.000000
25%     2853.000000  1959.000000   36141.75000    0.000000           1.000000
50%     5563.000000  1969.500000   52372.50000    0.000000           2.000000
75%     8584.250000  1977.000000   69293.25000    1.000000           3.000000
max    11191.000000  1995.000000  162397.00000    2.000000          15.000000

Из полученных данных можно сделать следующие выводы: минимальный год рождения клиента- 1899, максимальный – 1995, самый минимальный годовой доход семьи- 2447, максимальный – 162397, нет клиента, у которого больше двух детей, и максимальное количество покупок – 15.

 2.4 Оценка названий столбцов

In [6]:
# Вывести на экран названия столбцов с помощью df.columns. Выявить проблемы с названиями, если они есть. При необходимости переименовать столбцы. Если проблемы не обнаружены также дать пояснения.
df.columns

Index(['ID', 'Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome',
       'Dt_Customer', 'NumDealsPurchases'],
      dtype='object')

С помощью метода columns, выводятся названия столбцов

In [7]:
# переименование при необходимости
df=df.rename(columns={"NumDealsPurchases" : "Number_of_Purchases", "Dt_Customer":"Registration_Date"})

С помощью метода rename меняются названия у столбцов "Количество покупок" и "Дата регистрации клиента в компании" в связи с длинными и неудобными для восприятия названиями

### 3. Проверка пропусков

In [8]:
# Проверить данные на наличие пропусков и устранить их, если они есть (пропуски необходимо либо удалить, либо заменить каким-то значением).
print(df.isna().sum())

ID                      0
Year_Birth              0
Education               0
Marital_Status          0
Income                 12
Kidhome                 1
Registration_Date       1
Number_of_Purchases     1
dtype: int64


Для нахождения пропусков применяется метод isna(), а чтобы посчитать их количество, воспользуемся sum()

In [9]:
df=df.dropna(subset=['Income'])

Для удаления пустых строк в столбце Income используется метод dropna

### 4. Проверка дубликатов

#### Проверка явных дубликатов

In [10]:
df[df.duplicated()]

ID  Year_Birth   Education Marital_Status   Income  Kidhome  \
792  2853        1980  Graduation         Single  51766.0      1.0   
793  2853        1980  Graduation         Single  51766.0      1.0   
794  2853        1980  Graduation         Single  51766.0      1.0   
795  2853        1980  Graduation         Single  51766.0      1.0   

    Registration_Date  Number_of_Purchases  
792        11.03.2014                  2.0  
793        11.03.2014                  2.0  
794        11.03.2014                  2.0  
795        11.03.2014                  2.0

При проверке с помощью метода duplicated(), было выявлено 4 дубликата

In [11]:
# удалите дубликаты, если они есть
df=df.drop_duplicates().reset_index(drop=True)

С помощью метода drop_duplicates() удаляются дубликаты, а с помощью reset_index обновляется индексация. Также используется аргумент drop,чтобы не создавался столбец index

#### Проверка неявных дубликатов

In [12]:
columns_to_check=['Year_Birth','Education','Marital_Status','Kidhome','Number_of_Purchases']
for x in columns_to_check:
  print(x,":","\n", df[x].unique(),"\n")

Year_Birth : 
 [1957 1954 1965 1984 1981 1967 1971 1985 1974 1950 1976 1959 1952 1987
 1946 1980 1949 1982 1979 1951 1969 1989 1963 1970 1973 1943 1975 1968
 1964 1977 1978 1955 1966 1988 1948 1958 1972 1960 1983 1945 1991 1962
 1953 1956 1992 1961 1986 1990 1947 1899 1993 1994 1941 1944 1995] 

Education : 
 ['Graduation' 'PhD' 'Master' 'Basic'] 

Marital_Status : 
 ['Single' 'Together' 'Married' 'Divorced' 'SINGL' 'MARRIED' 'Widow'
 'Alone'] 

Kidhome : 
 [0. 1. 2.] 

Number_of_Purchases : 
 [ 3.  2.  1.  5.  4. 15.  7.  6.  9.  0.  8. 10. 13.] 



С помощью цикла for и метода для поиска уникальных значений df[x].unique(), выявляются дубликаты

In [13]:
# удалите дубликаты, если они есть
df["Marital_Status"]=df["Marital_Status"].replace("SINGL","Single")
df["Marital_Status"]=df["Marital_Status"].replace("MARRIED","Married")

Замена текущих значений новыми с помощью метода replace

### 5. Провека типов данных

In [14]:
print(df.isna().sum())

ID                     0
Year_Birth             0
Education              0
Marital_Status         0
Income                 0
Kidhome                0
Registration_Date      0
Number_of_Purchases    0
dtype: int64


Снова применяется метод isna() для проверки пропусков

In [15]:
# Проверьте типы данных, при необходимости измените типы данных, чтобы они соответствовали действительности.
df=df.astype({'Kidhome':'int64','Number_of_Purchases':'int64'})

Количество детей и количество покупок не может быть не целым значением, поэтому с помощью метода astype переводятся в целочисленный тип данных

In [16]:
df['Registration_Date']=pd.to_datetime(df['Registration_Date'],format='%d.%m.%Y')

С помощью метода to_datetime дата приводится в нужный формат(день, месяц, год)

### 6. Группировка данных

#### Задание 1

Группировка - семейный статус (marital_status) по количеству детей.


In [17]:
# выполните группировку согласно варианту
df.groupby(['Marital_Status','Kidhome'])['ID'].count()

Marital_Status  Kidhome
Alone           1            3
Divorced        0           52
                1           31
                2            2
Married         0          166
                1          125
                2           11
Single          0           97
                1           69
                2            3
Together        0          123
                1           66
                2            3
Widow           0           24
                1            5
Name: ID, dtype: int64

Для того, чтобы выполнить группировку семейного статуса по количеству детей используется метод groupby, а чтобы узнать количество используется count()

Из результата данной группировки можно увидеть, что больше всего людей не имеют детей,также,что самая большая группа: замужние/женатые клиенты (Married) без детей - 166 человек.

#### Задание 2

Группировка - тип образования по каждому семейному статусу
(marital_status). Создать датафрейм. Переименовать столбец с количеством в
“сount”. Отсортировать по убыванию столбца “count”.


In [18]:
# выполните группировку согласно варианту
df_group=df.groupby(['Education','Marital_Status'])['ID'].count().sort_values(ascending=False).reset_index(name='count')
df_group

Education Marital_Status  count
0   Graduation        Married    168
1   Graduation         Single     97
2   Graduation       Together     96
3          PhD        Married     68
4       Master        Married     57
5          PhD       Together     51
6   Graduation       Divorced     49
7       Master       Together     41
8          PhD         Single     40
9       Master         Single     31
10         PhD       Divorced     23
11  Graduation          Widow     21
12      Master       Divorced     12
13       Basic        Married      9
14         PhD          Widow      4
15      Master          Widow      4
16       Basic       Together      4
17       Basic       Divorced      1
18  Graduation          Alone      1
19       Basic         Single      1
20      Master          Alone      1
21         PhD          Alone      1

Здесь выполняется группировка типа образования по каждому семейному статусу, также происходит сортировка по убыванию столбца количества. Также используется метод groupby и count(), но помимо этого еще и sort_values(ascending=False), что говорит о том, что сортировка будет по убыванию. reset_index(name='count')- переименования столбца с количеством в "count"

Большинство клиентов с типом образования Graduation женаты. На втором месте по количеству одинокие клиенты также c типом образования Graduation. Типы образования PhD, Master, Graduation, Basic имеют по одному одинокому клиенту, тип образования Basic имеет одного разведенного клиента.

#### Задание 3

Сводная таблица (pivot_table) - среднее количество покупок для
каждого уровня образования. Отсортировать по убыванию. Округлить до двух
знаков.


In [19]:
# выполните сводную таблицу согласно варианту
pt=df.pivot_table(index='Education',values='Number_of_Purchases')
pt=pt.sort_values(by='Number_of_Purchases',ascending=False)
pt['Number_of_Purchases']=pt['Number_of_Purchases'].round(2)
pt

Number_of_Purchases
Education                      
Master                     2.53
PhD                        2.50
Graduation                 2.18
Basic                      1.40

В данном задании выполняется сводная таблица - среднее количество покупок для каждого уровня образования, отсортировать по убыванию и округлить до двух знаков после запятой. Для этого используется метод pivot_table для создания сводной таблицы и sort_values со значением False для сортировки по убыванию, а также round(2) для округления.

Самое большое среднее значение 2.53 для уровня образования Master, а самое маленькое 1.40 для Basic.

#### Задание 4

Сводная таблица (pivot_table) - средний доход семьи по семейному
положению - строки и уровню образованию - столбцы. Отсортировать по
возрастанию marital_status. Округлить до двух знаков.

In [20]:
# выполните сводную таблицу согласно варианту
pt2=df.pivot_table(index='Marital_Status',columns='Education',values='Income')
pt2=pt2.sort_values(by='Marital_Status',ascending=True)
pt2=pt2.fillna(0)
pt2=pt2.round(2)
pt2

Education          Basic  Graduation    Master       PhD
Marital_Status                                          
Alone               0.00    34176.00  61331.00  35860.00
Divorced         9548.00    57420.98  49405.83  53141.22
Married         22018.22    51078.81  55391.37  59391.00
Single          21063.00    50492.15  50492.45  53630.48
Together        22437.50    52922.06  51009.98  62801.33
Widow               0.00    55146.81  47164.00  65345.50

В этом задании была реализована сводная таблица - средний доход семьи по семейному положению(строки) и уровню образования(столбцы). Отсортировать по возрастанию Marital_Status, округлить до двух знаков. Для реализации такой таблицы используеттся также метод pivot_table, в котором указывается строки - index, columns - столбцы и values - значение среднего дохода. Происходит сортировка с помощью метода sort_values со значением True,что говорит о сортировки по возрастанию. Метод fillna(0) заменяет пустые значения на 0, а round(2) округляет до 2-х знаков после запятой.

Самый высокий доход имеют клиенты с типом образования PhD(кандидат наук) и семейным положением Widow. Нет клиентов с типом образования Basic и семейным положением Alone и Widow. Самый маленький средний доход у клиентов с типом образования Basic и семейным положением Divorced.

### Вывод


***`Сделать расширенный вывод. `***

Проведена обработка данных о клиентах магазина, максимальный год рождения клиента -1899, минимальный – 1995, минимальный годовой доход семьи – 2447, максимальный -162397, нет клиента, у которого больше двух детей, и максимальное количество покупок у клиента - 15. Можно сделать вывод,что самая большая группа: Замужние/женатые клиенты (Married) без детей - 166 человек. Подавляющее большинство клиентов во всех категориях семейного положения не имеют детей. Также большинство клиентов с типом образования Graduation женаты. На втором месте по количеству одинокие клиенты также c типом образования Graduation. Типы образования PhD, Master, Graduation, Basic имеют по одному одинокому клиенту, тип образования Basic имеет одного разведенного клиента.Самый высокий доход имеют клиенты с типом образования PhD(кандидат наук) и семейным положением Widow. Нет клиентов с типом образования Basic и семейным положением Alone и Widow. Самый маленький средний доход у клиентов с типом образования Basic и семейным положением Divorced.

В ходе лабораторной работы исправлены ошибочные данные. Применена группировка, сортировка и созданы сводные таблицы, в которых можно легко отыскать нужную информацию.


### Дополнительное задание

Создать столбец - возраст (расчетный). Создать столбец “Категория
возраста (Year_Birth)” (с помощью категоризации). Выделить минимум 3
категории (молодой, старый, средний), фильтрацию для уровня возраста
выбрать самостоятельно, аргументировать выбор. Создать сводную
таблицу: средний и медианный возраст по Marital_Status и категории
возраста.


In [21]:
# код выполнения задания
current_year = pd.Timestamp.now().year
df['Age'] = current_year - df['Year_Birth']

age_bins = [0, 35, 55, 100]
age_labels = ['Молодой', 'Средний', 'Пожилой']

df['Категория возраста (Year_Birth)'] = pd.cut(
    df['Age'],
    bins=age_bins,
    labels=age_labels,
    right=True
)

t=input('Введите категорию возраста ')
filtered_df = df[df['Категория возраста (Year_Birth)'] == t]
filtered_df.head(20)

Empty DataFrame
Columns: [ID, Year_Birth, Education, Marital_Status, Income, Kidhome, Registration_Date, Number_of_Purchases, Age, Категория возраста (Year_Birth)]
Index: []

Метод cut гарантирует, что все значения будут отнесены к соответствующим категориям согласно заданным границам, исключая субъективизм и ошибки ручной категоризации. Параметр right=True обеспечивает четкое определение принадлежности к интервалам. Выбор интерактивного подхода к фильтрации возрастных категорий через input()  позволяет гибко исследовать различные возрастные группы без необходимости модификации кода, что особенно ценно при работе с меняющимися исследовательскими данными.

In [22]:
pivot_table = pd.pivot_table(
    df,
    values='Age',
    index=['Marital_Status', 'Категория возраста (Year_Birth)'],
    aggfunc=['mean', 'median'],
    margins=True,
    margins_name='Все категории',
    observed=False
)

pivot_table.columns = ['Средний возраст', 'Медианный возраст']
pivot_table.round(1)

Средний возраст  \
Marital_Status Категория возраста (Year_Birth)                    
Alone          Средний                                     44.5   
               Пожилой                                     67.0   
Divorced       Средний                                     49.1   
               Пожилой                                     65.6   
Married        Молодой                                     34.2   
               Средний                                     47.5   
               Пожилой                                     66.0   
Single         Молодой                                     32.6   
               Средний                                     46.4   
               Пожилой                                     65.3   
Together       Молодой                                     33.0   
               Средний                                     47.5   
               Пожилой                                     67.3   
Widow          Средний                                     51.9   
               Пожилой                                     68.0   
Все категории                                              56.6   

                                                Медианный возраст  
Marital_Status Категория возраста (Year_Birth)                     
Alone          Средний                                       44.5  
               Пожилой                                       67.0  
Divorced       Средний                                       49.0  
               Пожилой                                       64.0  
Married        Молодой                                       34.5  
               Средний                                       49.0  
               Пожилой                                       65.0  
Single         Молодой                                       33.0  
               Средний                                       47.0  
               Пожилой                                       64.0  
Together       Молодой                                       33.5  
               Средний                                       48.0  
               Пожилой                                       67.0  
Widow          Средний                                       52.0  
               Пожилой                                       69.0  
Все категории                                                56.0

Была создана сводная таблица - средний и медианный возраст по Marital_Status и категории возраста

Создать столбец “Категория дохода (Income)” (с помощью категоризации).
Выделить минимум 3 категории (высокий, низкий, средний), фильтрацию
для уровня дохода выбрать самостоятельно, аргументировать выбор.
Создать сводные таблицы: средний и медианный NumDealsPurchases по
категории дохода и Education.

In [23]:
tercile_33 = df['Income'].quantile(0.33)
tercile_66 = df['Income'].quantile(0.66)
income_bins = [0, tercile_33, tercile_66, df['Income'].max() + 1]
income_labels = ['Низкий', 'Средний', 'Высокий']

df['Категория дохода (Income)'] = pd.cut(
    df['Income'],
    bins=income_bins,
    labels=income_labels,
    right=False
)
t=input('Введите категорию дохода ')
middle_income = df[df['Категория дохода (Income)'] == t]
middle_income.head(20)

Empty DataFrame
Columns: [ID, Year_Birth, Education, Marital_Status, Income, Kidhome, Registration_Date, Number_of_Purchases, Age, Категория возраста (Year_Birth), Категория дохода (Income)]
Index: []

Выбор интерактивного подхода к фильтрации уровня дохода через input()  позволяет гибко исследовать различный уровень дохода без необходимости модификации кода. Фильтрация основывается на статистически обоснованных категориях, созданных через терцильное разделение. Также был применен метод cut()

In [24]:
pivot_table = pd.pivot_table(
    df,
    values='Number_of_Purchases',
    index=['Категория дохода (Income)', 'Education'],
    aggfunc=['mean', 'median'],
    margins=True,
    margins_name='Все категории',
    observed=False
)

pivot_table.columns = ['Средние покупки ', 'Медианные покупки ']
pivot_table.round(2)

Средние покупки   Медианные покупки 
Категория дохода (Income) Education                                       
Низкий                    Basic                   1.40                 1.0
                          Graduation              2.16                 2.0
                          Master                  2.05                 2.0
                          PhD                     2.05                 2.0
Средний                   Graduation              2.85                 2.5
                          Master                  3.56                 3.0
                          PhD                     3.29                 3.0
Высокий                   Graduation              1.66                 1.0
                          Master                  1.31                 1.0
                          PhD                     2.04                 1.0
Все категории                                     2.31                 2.0

Была создана сводная таблица с помощью метода pivot_table - средний и медианный Number_of_Purchases по категории дохода и Education